services/github/github-auth-service.spec.js

```diff
@@ -25,17 +25,32 @@ describe('GithubAuthV3Service', function () {
     }
   }
 
-  it('forwards custom Accept header', async function () {
-    const sendAndCacheRequestWithCallbacks = sinon.stub().returns(
+  class ScopedDummyGithubAuthV3Service extends DummyGithubAuthV3Service {
+    constructor(context, config) {
+      super(context, config, { needsPackageScope: true })
```

Comment:
This partly relates to the conversation we were having yesterday about integrations potentially needing >1 scope. I'd tend to think of this as being `scopes: ['read:package']`. I think it also probably makes sense to keep the terminology at the service level (i.e: code a casual contributor is going to write) using the documented scope names https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps#available-scopes

Reply:
I originally started out doing what you're suggesting, but here's the main reasons I decided to pivot to what's here now:

- using named boolean properties doesn't prevent the ability for a service class to request multiple scopes, it's just a different "how" (e.g. `{ needsPackageScope: true, needsReadOrgScope: true }`), plus it also gives us the ability to define our own grouping scopings that can be mapped directly to a corresponding pool (e.g. `{ needsPackageAndUserScope: true }`)
- we're working with an entirely constrained, and small universe of possibilities which are all known at build time. As such there's no reason to have to iterate over collections and do string comparisons on every request at runtime

I acknowledge and appreciate that there's a lower barrier for someone _writing_ a service class in isolation to be able to just copy in the literal scope value from the GitHub docs into a `scopes` array, and that using named boolean properties introduces a "Shields-ism" that's not immediately obvious to new contributors.

However, I don't view that as a significant problem nor hurdle. New GitHub badges are not exactly a common occurrence, and I think we can cover this pretty easily via the review process and documentation, especially since a net-new badge that requires a net new scope/scope combination is going to require touching some core code (at least in the api provider) anyway.